contracts/transfer/src/state.rs

```diff
@@ -198,45 +170,54 @@ impl TransferState {
         &mut self,
         tx: Transaction,
     ) -> Result<Vec<u8>, ContractError> {
-        //  1. α ∈ R
-        if !self.root_exists(&tx.anchor) {
-            panic!("Anchor not found in the state!");
+        let tx_skeleton = tx.payload().tx_skeleton();
+
+        // panic if the transaction fails basic validation
+        assert!(tx.is_valid(), "Transaction is invalid");
+
+        // panic if the root is invalid
+        if !self.root_exists(&tx_skeleton.root) {
+            panic!("Root not found in the state!");
         }
 
-        //  2. ν[] !∈ Nullifiers
-        if self.any_nullifier_exists(&tx.nullifiers) {
+        // panic if any of the given nullifiers already exist
+        if self.any_nullifier_exists(&tx_skeleton.nullifiers) {
             panic!("A provided nullifier already exists!");
         }
 
-        //  3. Nullifiers.append(ν[])
-        self.nullifiers.extend(&tx.nullifiers);
-
-        //  4. if |C|=0 then set C ← (0,0,0)
-        //  Crossover is received as option
+        // append the nullifiers to the nullifiers set
+        self.nullifiers.extend(&tx_skeleton.nullifiers);
 
-        //  5. N↦.append((No.R[], No.pk[])
-        //  6. Notes.append(No[])
-        let block_height = rusk_abi::block_height();
-        self.tree.extend_notes(block_height, tx.outputs.clone());
-
-        //  7. g_l < 2^64
-        //  8. g_pmin < g_p
-        //  9. fee ← g_l ⋅ g_p
-        // 10. verify(α, ν[], C.c, No.c[], fee)
+        // verify the phoenix-circuit
         if !verify_tx_proof(&tx) {
             panic!("Invalid transaction proof!");
         }
 
-        // 11. if ∣k∣≠0 then call(k)
-        self.var_crossover = tx.crossover;
-        self.var_crossover_addr.replace(*tx.fee.stealth_address());
+        // append the output notes to the phoenix-notes tree
+        let block_height = rusk_abi::block_height();
+        self.tree
+            .extend_notes(block_height, tx_skeleton.outputs.clone());
+
+        // place the contract deposit on the state
+        if tx.payload().deposit {
+            let contract = match tx.payload().contract_call() {
+                Some(call) => ContractId::from_bytes(call.contract),
+                None => {
+                    panic!("There needs to be a contract call when depositing funds");
```

Comment:
Is this true? Could we not allow it and then, if not picked up by a contract, simply return it? That's what should happen anyway when a contract doesn't pick it up regardless.

Reply:
We could, my idea was to discard those `Transaction`s without nullifying the input-`Note`s

Reply:
The assumption is that contract call calls 'deposit' right? What if the contract call doesn't call 'deposit'?